### ISSUE TITLE
Discussion: Directives inside CodeBlocks

### ISSUE THREAD

**Issue by NTaylorMullen** [aspnet/Razor#61]
@pranavkm and I had a talk about directives inside codeblocks on whether we should continue allowing them.  For example:

```
@{
    @inherits Foo
}
```

Is valid in MVC5.  Should we continue allowing this in vNext?

It blocks scenarios such as:

```
@{
    var inherits = true;
    @inherits Foo
}
```

Because `@inherits` is a keyword it blocks the C# keyword inherits from being used.  Not only that but it doesn't make complete sense to add an inherits statement within a codeblock to begin with.  We should discuss if we want to allow this behavior or not.

**Comment by GuardRex** [comment]
Just ran into a related issue along the lines of how Razor directives are specified. I'm using `gulp-minify-html` on "debug" views and sending the minified views into the Views folder. As it stands, razor directives aren't going to play well with the minifier without some refactoring to the plugin. For example, starting with this ...
```
@model MyApp.Models.Visitor
@tagHelperPrefix "th:"
@addTagHelper "*, Microsoft.AspNet.Mvc.TagHelpers"
@{
    Layout = "/Views/Shared/_ContactLayout.cshtml";
    ViewBag.Title = "Contact";
    ViewBag.Description = "Please feel free to contact us with the contact webpage.";
}
<div id="contact">
... more markup here ...
```
is minfied to:
```
@model paccharter.Models.Visitor @tagHelperPrefix "th:" @addTagHelper "*, Microsoft.AspNet.Mvc.TagHelpers" @{ Layout = "/Views/Shared/_ContactLayout.cshtml"; ViewBag.Title = "Contact"; ViewBag.Description = "Please feel free to contact us with the contact webpage."; }<div id=contact>... more markup here ...
```
... which breaks of course.

It would be cool if the Razor directives either had a line terminator so that minfication of the view didn't break them or if they could indeed be put into the codeblock to protect them from minification.

**Comment by GuardRex** [comment]
Followup to my Apr 27 comment: I ended up using a `gulp-inject-string` task after my `gulp-minify-html` task to get the Razor directives on their own lines at the tops of my views after they are minified into my Views folder.

Gulp plugins:
```
var gulp = require('gulp'),
    minifyHTML = require('gulp-minify-html'),
    inject = require('gulp-inject-string');
```
So starting with this "debug" view:
```
@{
    Layout = "/Views/Shared/_ContactLayout.cshtml";
    ViewBag.Title = "Contact";
    ViewBag.Description = "Please feel free to contact us with the contact webpage.";
}
<div id="contact">
... more markup here ...
```
I use this:
```
gulp.task('processContactViews', function () {
    gulp.src('Views_debug/Contact/*.cshtml')
        .pipe(minifyHTML({ quotes: true }))
        .pipe(inject.prepend('@model wpuc.Models.Visitor\n@tagHelperPrefix "th:"\n@addTagHelper "*, Microsoft.AspNet.Mvc.TagHelpers"\n'))
        .pipe(gulp.dest('Views/Contact'));
});
```
which results in the Razor parser receiving this:
```
@model paccharter.Models.Visitor
@tagHelperPrefix "th:"
@addTagHelper "*, Microsoft.AspNet.Mvc.TagHelpers"
@{ Layout = "/Views/Shared/_ContactLayout.cshtml"; ViewBag.Title = "Contact"; ViewBag.Description = "Please feel free to contact us with the contact webpage."; }<div id=contact>... more markup here ...
```
which it's perfectly happy to parse, and I get HTML minification.

I mention this here, because it looks like the team will revisit this issue from the Backlog at some point in the future.

It's a pain to minify Razor views (as also pointed out by @RehanSaeed in https://github.com/aspnet/Razor/issues/423). If the team gets back to thinking more about the way Razor directives are specified (in a codeblock as this issue suggests or not), I'd like to plug the idea that they need a line termination character and a mod to the Razor parser to tease them out of minified markup using that character. It seems like an oversight that when Razor was invented that the engineers didn't think that relying upon line feeds in markup (where they have never been required) was a bit of a regression for markup generally. 

I made Razor work with HTML minification, but my hack still sucks because (I think) not having the directives in the debug view is preventing Intellisense that I would get when working with the markup (e.g., view model binding). I also dislike having to have a Views_debug copy of all of my markup. If anyone has a better approach for minifying Razor views, I'm all :ear:'s.
